### PR TITLE
[10.0][FIX] auth_totp: Various issues

### PR DESCRIPTION
**Pull request by obulkin**
* Restructure controller and ``res.users`` logic to prevent RPC authentication for users with MFA enabled and add support for multiple simultaneous MFA sessions
* Switch trusted device cookies from using the DB secret to user-level secret keys, thereby increasing security
* Remove MFA login tokens and trusted device model, which are now redundant
* Add migration logic that generates a trusted device cookie key for every user with MFA enabled and cleans up device model ``ir`` records to prevent warnings
* Update unit tests and remainder of module accordingly

This PR addresses the issues mentioned in https://github.com/OCA/server-tools/issues/1041.

### CODE REVIEW

File: auth_totp/controllers/main.py
Comment (lasley) [inline auth_totp/controllers/main.py]: The `in` after `login` is superfluous. Or maybe there's a word missing?

File: auth_totp/migrations/10.0.2.0.0/post-migrate.py
Comment (lasley) [inline auth_totp/migrations/10.0.2.0.0/post-migrate.py]: Typically you want to import only the namespace(s) you're using (with obvious exceptions if it starts getting out of hand). So in this case, `from uuid import uuid4` would be better.

File: auth_totp/models/res_users.py
Comment (lasley) [inline auth_totp/models/res_users.py]: `from uuid import uuid4`

File: auth_totp/models/res_users.py
Comment (lasley) [inline auth_totp/models/res_users.py]: `record.trusted_device_cookie_key = record.mfa_enabled and uuid.uuid4() or False`

^--I'm almost sure that'll fit on one line once the nesting is reduced. Otherwise, just use an intermediary var:

```
key = ...
record.tru... = ...
```

File: auth_totp/models/res_users.py
Comment (lasley) [inline auth_totp/models/res_users.py]: Should end your docblocks with periods ([PEP-0257](https://www.python.org/dev/peps/pep-0257/#one-line-docstrings) Single line docstring note 4). 

The bullet points should also have punctuation, because they are sentences that complete the introductory stem (`Overview:`)

File: auth_totp/migrations/10.0.2.0.0/post-migrate.py
Comment (lasley) [inline auth_totp/migrations/10.0.2.0.0/post-migrate.py]: Period at end of docblock sentence

File: auth_totp/models/res_users.py
Comment (lasley) [inline auth_totp/models/res_users.py]: `mfa_login_active` string key seems pretty integral to things, and is repeated in a lot of spots. Typically I try to make strings into constants once I've used them 3-4 times (not counting tests). Do we have a good place to do this? Maybe in MFA model?

**Review by lasley — CHANGES_REQUESTED** [review]
Thanks @obulkin - I like this design much better. Comments inline

File: auth_totp/models/res_users.py
Comment (fmdl) [inline auth_totp/models/res_users.py]: `user = self.sudo().browse(self.env.uid)`

**Comment by fmdl** [comment]
It could great that a basic user cannot desactivate the MFA, for more security.
```
@multi
def write(self, vals):
    if 'mfa_enable' in vals.keys():
         if vals['mfa_enable'] == False and not self.user_has_group('base.group_user_root'):
              raise UserError(_('You have not allowed to desactivate MFA'))
```

**Comment by lasley** [comment]
> It could great that a basic user cannot desactivate the MFA, for more security.

This would be a security consideration that a company administrator would need to make as an option IMO. At the moment, there aren't really any options for things like that, so this would expand the scope of this bug fix quite drastically I think. 

@obulkin - we have an internal ticket for it already I think, but would be good to add to roadmap as well.

File: auth_totp/models/res_users.py
Comment (lasley) [reply auth_totp/models/res_users.py]: Why not `self.env.user.sudo()`?

File: auth_totp/controllers/main.py
Comment (obulkin) [reply auth_totp/controllers/main.py]: The sentence is referring to the ``login`` value in the ``session`` dictionary.

File: auth_totp/models/res_users.py
Comment (obulkin) [reply auth_totp/models/res_users.py]: This doesn't really seem like a substantive improvement, especially since two lines will be needed, so I'm leaving it as is.

File: auth_totp/models/res_users.py
Comment (obulkin) [reply auth_totp/models/res_users.py]: Actually, it turns out that ``env.user`` already [bypasses access restrictions](https://github.com/odoo/odoo/blob/10.0/odoo/api.py#L796), so I'll just drop these lines entirely and use that as needed. Thanks for the push in the right direction.

File: auth_totp/models/res_users.py
Comment (obulkin) [reply auth_totp/models/res_users.py]: This key is only used in two places outside of tests, and any constant used to replace it will need to be just as verbose, so I don't think anything is gained by doing that.

**Comment by obulkin** [comment]
I've updated this PR to address the comments made so far.

@lasley - There is already an item in the README roadmap that effectively covers this request to prevent MFA from being disabled:

> * Add company-level settings for forcing all users to enable MFA and disabling 
  the trusted device option

I agree that this would be a substantial new feature and doesn't really belong in a bugfix PR.

**Comment by obulkin** [comment]
Travis is currently failing due to unrelated issues with the ``sentry`` module.

**Review by lasley — APPROVED** [review]
Changes look good, thanks @obulkin. Confirmed Travis unrelated

**Comment by fmdl** [comment]
Hi guys, when it is merge ?

**Comment by lasley** [comment]
@fmdl - [Two review approvals are required for merge](https://github.com/OCA/maintainer-tools/blob/master/CONTRIBUTING.md#please-respect-a-few-basic-rules). There is only one here (mine). If you would like to see the merge faster, please review the work and approve if satisfactory.

**Comment by fmdl** [comment]
Dear @lasley, 

Your update doesn't prevent an RPC acces.

This `uid = common.authenticate(db, username, password, {})` return `False`, for authentification it is works.

But if you know the uid (it not hard to find) and the password, RPC access is allowed:
`products = models.execute_kw(db, 1, password,'product.product', 'search', [[]], {'limit':1})`

The best way is to put the password in a new field

**Review by fmdl — CHANGES_REQUESTED** [review]
Your update doesn't prevent an RPC acces.

This uid = common.authenticate(db, username, password, {}) return False, for authentification it is works.

But if you know the uid (it not hard to find) and the password, RPC access is allowed:
products = models.execute_kw(db, 1, password,'product.product', 'search', [[]], {'limit':1})

**Comment by obulkin** [comment]
@fmdl - Nice find. This only happens if there is a successful web login for the MFA user prior to the RPC request and is caused by a caching mechanism that skips the call to ``check_credentials``: https://github.com/odoo/odoo/blob/10.0/odoo/addons/base/res/res_users.py#L503. Changing the password field would not help, but I do have a solution in mind and will soon update this PR with the fix.

**Comment by fmdl** [comment]
It seems there are a bug in odoo, see here : https://github.com/odoo/odoo/pull/22212

**Review by fmdl — APPROVED** [review]

**Comment by obulkin** [comment]
@fmdl - I've pushed up a commit that mostly closes this RPC security hole, but it does have a limitation. Because the cache is implemented as a class attribute, it is not synced up across threads/processes. This means that when a user first enables MFA, there is no way to ensure that the cache states of every thread/process reflect this change until the server has been restarted (in the single-threaded, single-process scenario, just logging out and logging back in is enough). The only way around this involves accessing the DB in ``check`` and would offset the performance benefits of the caching, so we've decided to leave it as a known issue for now.

**Review by lasley — APPROVED** [review]
I'm good with this, thanks @obulkin

**Comment by obulkin** [comment]
@pedrobaeza Just wanted to pop in here to recommend merging this PR. It closes some significant security holes in the module and simplifies a lot of the logic, which should make it easier for others to maintain and upgrade the code.

**Review by MiquelRForgeFlow — APPROVED** [review]
I'm interested in migrating this module to v11

**Pull request merged**
